src/context-selection/org-unit-selector-bar-item/use-load-offline-levels.js

```diff
@@ -0,0 +1,43 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useEffect, useState } from 'react'
+import loadOfflineLevel from './load-offline-level.js'
+import useOfflineLevelsToLoad from './use-offline-levels-to-load.js'
+import useOrganisationUnitLevels from './use-organisation-unit-levels.js'
+
+/**
+ * As the service worker caches request responses,
+ * it should suffice to simply perform all request the org unit tree would
+ * perform as well in advance
+ */
+export default function useLoadOfflineLevels() {
+    const [loading, setLoading] = useState(true)
+    const [error, setError] = useState(undefined)
+    const [data, setData] = useState(false)
+
+    const dataEngine = useDataEngine()
+    const organisationUnitLevels = useOrganisationUnitLevels()
+    const offlineLevelsToLoad = useOfflineLevelsToLoad(organisationUnitLevels)
+    const { data: offlineLevelsToLoadData } = offlineLevelsToLoad
+
+    useEffect(() => {
+        // Can't pass async function to useEffect
+        offlineLevelsToLoadData &&
+            (async () => {
+                await Promise.all(
+                    offlineLevelsToLoadData.map((offlineLevelToLoad) => {
+                        const payload = {
+                            dataEngine,
+                            ...offlineLevelToLoad,
+                        }
+
+                        return loadOfflineLevel(payload)
+                    })
+                )
+                    .then(() => setData(true))
+                    .catch(setError)
+                    .finally(() => setLoading(false))
+            })()
+    }, [dataEngine, offlineLevelsToLoadData, setLoading, setData])
```

Comment:
```suggestion
    const loadOfflineLevels = async () => {
        setLoading(true)
        try {
            await Promise.all(
                offlineLevelsToLoadData.map((offlineLevelToLoad) => {
                    const payload = {
                        dataEngine,
                        ...offlineLevelToLoad,
                    }

                    return loadOfflineLevel(payload)
                })
            )
            setData(true)
        } catch (error) {
            setError(error)
        }
        setLoading(false)
    }

    useEffect(() => {
        if (offlineLevelsToLoadData) {
            loadOfflineLevels()
        }
    }, [dataEngine, offlineLevelsToLoadData, setLoading, setData])
```

Reply:
I've done it slightly differently, I moved the function out of the hook